Sources/ApolloCodegenLib/Frontend/GraphQLSchema.swift

```diff
@@ -0,0 +1,119 @@
+  }
+    
+  func isSubType(abstractType: GraphQLAbstractType, maybeSubType: GraphQLNamedType) throws -> Bool {
+    return try invokeMethod("isSubType", with: abstractType, maybeSubType)
+  }
+}
+
+public class GraphQLNamedType: JavaScriptObject {
+  private(set) lazy var name: String = self["name"]
+}
+
+extension GraphQLNamedType: CustomDebugStringConvertible {
+  public var debugDescription: String {
+    return "<\(type(of: self)): \(jsValue.toString() ?? "?")>"
+  }
+}
+
+public class GraphQLScalarType: GraphQLNamedType {
+  private(set) lazy var description: String? = self["description"]
+  
+  private(set) lazy var specifiedByURL: String? = self["specifiedByUrl"]
+}
+
+public class GraphQLEnumType: GraphQLNamedType {
+  private(set) lazy var description: String? = self["description"]
+  
+  private(set) lazy var values: [GraphQLEnumValue] = try! invokeMethod("getValues")
+}
+
+public class GraphQLEnumValue: JavaScriptObject {
+  private(set) lazy var name: String = self["name"]
+  
+  private(set) lazy var description: String? = self["description"]
+    
+  private(set) lazy var deprecationReason: String? = self["deprecationReason"]
+}
+
+public class GraphQLInputObjectType: GraphQLNamedType {
+  private(set) lazy var description: String? = self["description"]
+  
+  private(set) lazy var fields: [String: GraphQLInputField] = try! invokeMethod("getFields")
+}
+
+public class GraphQLInputField: JavaScriptObject {
+  private(set) lazy var name: String = self["name"]
+  
+  private(set) lazy var type: GraphQLType = self["type"]
+  
+  private(set) lazy var description: String? = self["description"]
+  
+  private(set) lazy var defaultValue: Any? = self["defaultValue"]
+    
+  private(set) lazy var deprecationReason: String? = self["deprecationReason"]
+}
+
+public class GraphQLCompositeType: GraphQLNamedType {
+}
+
+public class GraphQLObjectType: GraphQLCompositeType {
+  private(set) lazy var description: String? = self["description"]
+  
+  private(set) lazy var fields: [String: GraphQLField] = try! invokeMethod("getFields")
+  
+  private(set) lazy var interfaces: [GraphQLInterfaceType] = try! invokeMethod("getInterfaces")
+}
+
+public class GraphQLAbstractType: GraphQLCompositeType {
+}
+
+public class GraphQLInterfaceType: GraphQLAbstractType {
+  private(set) lazy var description: String? = self["description"]
+  
+  private(set) lazy var deprecationReason: String? = self["deprecationReason"]
+  
+  private(set) lazy var fields: [String: GraphQLField] = try! invokeMethod("getFields")
+  
+  private(set) lazy var interfaces: [GraphQLInterfaceType] = try! invokeMethod("getInterfaces")
+}
+
+public class GraphQLUnionType: GraphQLAbstractType {
+  private(set) lazy var types: [GraphQLObjectType] = try! invokeMethod("getTypes")
+}
+
+public class GraphQLField: JavaScriptObject {
+  private(set) lazy var name: String = self["name"]
+  
+  private(set) lazy var type: GraphQLType = self["type"]
```

Comment:
what happens here if `self["type"]` doesn't convert to `GraphQLType`?

Reply:
That would trigger one of our own preconditions, if it isn't a JavaScript object or if it is not of the expected type. I realize preconditions sound dangerous, but I think they are the right choice here because this signals a programmer error and not something that should happen under normal circumstances. We also don't have much of a choice because Swift getters can't be throwing.

Reply:
Right - I'm mostly thinking about the readability of the errors we throw here. What will a user see that will allow them to do anything about this?

Reply:
Users can't do anything about this. If this happens, there is something wrong with our JavaScript or our mapping. That should be caught in development.

Reply:
I think this gets to my same question as above with the raw `"undefined"` string coming back - it does seem like something we should be able to `fatalError` on internally for a clearer idea of what's going on.

Reply:
That's exactly what we're doing with preconditions: they test explicit conditions and generate specific errors.